rust/kernel/src/lib.rs

```diff
@@ -66,25 +63,7 @@ static ALLOCATOR: allocator::KernelAllocator = allocator::KernelAllocator;
 /// Attempts to allocate memory for `value` using the global allocator. On success, `value` is
 /// moved into it and returned to the caller wrapped in a `Box`.
 pub fn try_alloc<T>(value: T) -> KernelResult<Box<T>> {
-    let layout = Layout::new::<MaybeUninit<T>>();
-    let ptr: NonNull<MaybeUninit<T>> = if layout.size() == 0 {
-        NonNull::dangling()
-    // SAFETY: We checked that the layout size is nonzero.
-    } else if let Some(nn) = NonNull::new(unsafe { alloc::alloc::alloc(layout) }) {
-        nn.cast()
-    } else {
-        return Err(Error::ENOMEM);
-    };
-
-    unsafe {
-        // SAFETY: `ptr` was just allocated and isn't used afterwards.
-        let mut b = Box::from_raw(ptr.as_ptr());
-        // SAFETY: The pointer is valid for write and is properly aligned. The dangling pointer
-        // case is only when the size of the value is zero; writing zero bytes to it is allowed.
-        b.as_mut_ptr().write(value);
-        // SAFETY: The value was initialised in the call above.
-        Ok(Box::from_raw(Box::into_raw(b) as *mut T))
-    }
+    Box::try_new(value).map_err(|_| Error::ENOMEM)
```

Comment:
I suggest we implement `From<TryReserveError>` for `Error` -- this way we can avoid the `map_err` call.

BTW, Manish implemented `try_new` for Box, Arc, Rc in response to an internal discussion we had (see this [comment](https://github.com/rust-lang/rust/issues/48043#issuecomment-748137026)) -- it's nice to see support from the rust core team.

Reply:
@wedsonaf Thanks a lot for moving it forward!

Reply:
Oooh yes, that sounds quite good. I'll send a follow up PR. That shoudl make the code sufficiently trivial as to allow me to just delete this function.

Most excellent -- by chance I had DM'd Manish this PR as a thank you, so I suppose that brings this all full circle!